pkg/network/tracer/process_cache.go

```diff
@@ -0,0 +1,299 @@
+	lru "github.com/hashicorp/golang-lru"
+)
+
+var defaultFilteredEnvs = []string{
+	"DD_ENV",
+	"DD_VERSION",
+	"DD_SERVICE",
+}
+
+const (
+	maxProcessQueueLen = 100
+	// maxProcessListSize is the max size of a processList
+	maxProcessListSize = 5
+)
+
+type process struct {
+	Pid         uint32
+	Envs        map[string]string
+	ContainerID string
+	StartTime   int64
+}
+
+type processList []*process
+
+type _processCacheStats struct {
+	cacheEvicts   uint64 `stats:"atomic"`
+	cacheLength   uint64 `stats:"atomic"`
+	eventsDropped uint64 `stats:"atomic"`
+	eventsSkipped uint64 `stats:"atomic"`
+
+	reporter stats.Reporter
+}
+
+type processCache struct {
+	sync.Mutex
+
+	// cache of pid -> list of processes holds a list of processes
+	// with the same pid but differing start times up to a max of
+	// maxProcessListSize. this is used to determine the closest
+	// match to a connection's tiimestamp
+	cacheByPid map[uint32]processList
+	// lru cache; keyed by (pid, start time)
+	cache *lru.Cache
+	// filteredEnvs contains environment variable names
+	// that a process in the cache must have; empty filteredEnvs
+	// means no filter, and any process can be inserted the cache
+	filteredEnvs []string
+
+	in      chan *process
+	stopped chan struct{}
+	stop    sync.Once
+
+	stats *_processCacheStats
+}
+
+type processCacheKey struct {
+	pid       uint32
+	startTime int64
+}
+
+func newProcessCache(maxProcs int, filteredEnvs []string) (*processCache, error) {
+	pc := &processCache{
+		filteredEnvs: filteredEnvs,
+		cacheByPid:   map[uint32]processList{},
+		in:           make(chan *process, maxProcessQueueLen),
+		stopped:      make(chan struct{}),
+		stats:        &_processCacheStats{},
+	}
+	var err error
+	pc.cache, err = lru.NewWithEvict(maxProcs, func(key, value interface{}) {
+		p := value.(*process)
+		pl, _ := pc.cacheByPid[p.Pid]
+		if pl = pl.remove(p); len(pl) == 0 {
+			delete(pc.cacheByPid, p.Pid)
+			return
+		}
+
+		pc.cacheByPid[p.Pid] = pl
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if pc.stats.reporter, err = stats.NewReporter(pc.stats); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for {
+			select {
+			case <-pc.stopped:
+				return
+			case p := <-pc.in:
+				pc.add(p)
+			}
+		}
+	}()
+
+	return pc, nil
+}
+
+func (pc *processCache) handleProcessEvent(entry *smodel.ProcessCacheEntry) {
+
+	select {
+	case <-pc.stopped:
+		return
+	default:
+	}
+
+	p := pc.processEvent(entry)
+	if p == nil {
+		atomic.AddUint64(&pc.stats.eventsSkipped, 1)
+		return
+	}
+
+	select {
+	case pc.in <- p:
+	default:
+		// dropped
+		atomic.AddUint64(&pc.stats.eventsDropped, 1)
+	}
+}
+
+func (pc *processCache) processEvent(entry *smodel.ProcessCacheEntry) *process {
+	var envs map[string]string
+	if entry.EnvsEntry != nil {
+		for _, v := range entry.EnvsEntry.Values {
+			kv := strings.SplitN(v, "=", 2)
+			k := kv[0]
+			found := len(pc.filteredEnvs) == 0
+			for _, r := range pc.filteredEnvs {
```

Comment:
Would it be better to use a map here instead of iterating over the `filteredEvents` slice?